### PR TITLE
Ethdoc plugin

### PR DESCRIPTION
**Pull request by joeizang**
This PR generates documentation from compiled contracts using openzeplin docgen libraries @bunsenstraat adapted for remix ide. Follow these steps to test:
- Checkout this branch locally
- using the NX VsCode plugin, run both docviewer and remixdocgen plugins
- add local iframe plugins for both remixdocgen (sidepanel on localhost:6003) and docviewer (mainPanel on localhost:7003).
- compile any contract and switch to docgen plugin and click on generate (note that this sidepanel UI is very temporary for now) and the docviewer will display the generated documentation.

The docviewer handles the generated md files  and displays them. You will find the docs inside the docs folder in File Explorer

fix https://github.com/ethereum/remix-project/issues/3462

### CODE REVIEW

**Comment by yann300** [comment]
please update the naming of the plugin (the folder name too), this should be `doc-viewer` and `doc-gen`

File: apps/doc-gen/src/app/docgen-client.ts [hunk 1]
```diff
@@ -0,0 +1,82 @@
+import { PluginClient } from '@remixproject/plugin'
+import { CompilationResult, SourceWithTarget } from '@remixproject/plugin-api'
+import { createClient } from '@remixproject/plugin-webview'
+import EventEmitter from 'events'
+import { Config, defaults } from './docgen/config'
+import { Build, buildSite } from './docgen/site'
+import { loadTemplates } from './docgen/templates'
+import {SolcInput, SolcOutput} from 'solidity-ast/solc'
```
Comment (LianaHus) [inline apps/doc-gen/src/app/docgen-client.ts:8]: spaces around missing

File: apps/doc-gen/src/app/docgen-client.ts [hunk 2]
```diff
@@ -0,0 +1,82 @@
...
+      const output: SolcOutput = {
+        sources: data.sources as any
+      }
+      this.build = {
+        input: input,
+        output: output
+      }
+      this.fileName = fileName
+      this.eventEmitter.emit('compilationFinished', this.build, fileName)
+
```
Comment (LianaHus) [inline apps/doc-gen/src/app/docgen-client.ts:50]: extra line

File: apps/doc-gen/src/app/docgen/site.ts [hunk 3]
```diff
@@ -0,0 +1,138 @@
...
+    }
+  }
+
+  return {
+    items,
+    pages: Object.entries(pages).map(([id, pageItems]) => ({ id, items: pageItems })),
+  };
+}
+
+function defineContext(item: DocItem, build: BuildContext, file: SourceUnit, page?: string, contract?: ContractDefinition): DocItemWithContext {
```
Comment (LianaHus) [inline apps/doc-gen/src/app/docgen/site.ts:109]: you need space after function names

File: apps/doc-gen/src/app/docgen/site.ts [hunk 4]
```diff
@@ -0,0 +1,138 @@
...
+}
+
+function defineProperties(item: DocItemWithContext, properties: Properties) {
+  for (const [prop, fn] of Object.entries(properties)) {
+    const original: unknown = (item as any)[prop];
+    defineGetterMemoized(item as any, prop, () => fn(item.__item_context, original));
+  }
+}
+
+function assignIfIncludedSource(
```
Comment (LianaHus) [inline apps/doc-gen/src/app/docgen/site.ts:122]: in this doc as well

File: apps/doc-gen/src/app/docgen/templates.ts [hunk 5]
```diff
@@ -0,0 +1,95 @@
...
+}
+
+/**
+ * Reads all built-in themes into an object. Partials will always be found in
+ * src/themes, whereas helpers may instead be found in dist/themes if TypeScript
+ * can't be imported directly.
+ */
+async function readThemes(): Promise<Record<string, Required<Templates>>> {
+  const themes: Record<string, Required<Templates>> = {};
+
```
Comment (LianaHus) [inline apps/doc-gen/src/app/docgen/templates.ts:88]: extra lines

File: apps/doc-gen/src/index.html [hunk 6]
```diff
@@ -0,0 +1,14 @@
+<!DOCTYPE html>
+<html lang="en">
+  <head>
+    <meta charset="utf-8" />
+    <title>Remix Docgen</title>
+    <base href="/" />
+
```
Comment (LianaHus) [inline apps/doc-gen/src/index.html:7]: extra line

File: apps/doc-gen/webpack.config.js [hunk 7]
```diff
@@ -0,0 +1,85 @@
...
+
+  // add externals
+  config.externals = {
+    ...config.externals,
+    solc: 'solc',
+  }
+
+  // add public path
+  config.output.publicPath = '/'
+
```
Comment (LianaHus) [inline apps/doc-gen/webpack.config.js:42]: extra lines

File: apps/doc-viewer/src/index.html [hunk 8]
```diff
@@ -0,0 +1,14 @@
+<!DOCTYPE html>
+<html lang="en">
+  <head>
+    <meta charset="utf-8" />
+    <title>Doc Viewer</title>
+    <base href="/" />
+
```
Comment (LianaHus) [inline apps/doc-viewer/src/index.html:7]: extra line

File: apps/doc-viewer/webpack.config.js [hunk 9]
```diff
@@ -0,0 +1,85 @@
...
+
+  // add externals
+  config.externals = {
+    ...config.externals,
+    solc: 'solc',
+  }
+
+  // add public path
+  config.output.publicPath = '/'
+
```
Comment (LianaHus) [inline apps/doc-viewer/webpack.config.js:42]: extra lines

File: apps/doc-viewer/webpack.config.js [hunk 10]
```diff
@@ -0,0 +1,85 @@
...
+  // souce-map loader
+  config.module.rules.push({
+    test: /\.js$/,
+    use: ["source-map-loader"],
+    enforce: "pre"
+  })
+
+  config.ignoreWarnings = [/Failed to parse source map/] // ignore source-map-loader warnings
+
```
Comment (LianaHus) [inline apps/doc-viewer/webpack.config.js:65]: extra line

File: apps/doc-gen/src/app/docgen/common/properties.ts [hunk 11]
```diff
@@ -0,0 +1,138 @@
...
+import { DocItemContext, DOC_ITEM_CONTEXT } from '../site';
+import { mapValues } from '../utils/map-values';
+import { DocItem, docItemTypes } from '../doc-item';
+import { formatVariable } from './helpers';
+import { PropertyGetter } from '../templates';
+import { itemType } from '../utils/item-type';
+
+type TypeDefinition = StructDefinition | EnumDefinition | UserDefinedValueTypeDefinition;
+
+export function type({ item }: DocItemContext): string {
```
Comment (LianaHus) [inline apps/doc-gen/src/app/docgen/common/properties.ts:13]: please add space after func names in the file

**Comment by LianaHus** [comment]
i was not able to make it work. 
- both plugins are in plugin manager (I guess they shouldn't be there)
- after activation one is empty the other one I don;t know what happens, I don't see it.
- there is no Generate Docs in right click menu

File: apps/doc-gen/src/app/App.tsx [hunk 12]
```diff
@@ -0,0 +1,43 @@
...
+    }
+    watchThemeSwitch()
+  }, [])
+
+  return (
+    <div className="p-3">
+      <h1>Remix Docgen</h1>
+      {fileName && <h4>File: {fileName.split('/')[1].split('.')[0].concat('.sol')}</h4>}
+      {hasBuild && <button className="btn btn-primary btn-block mt-4 rounded" onClick={() => client.generateDocs()}>Generate doc</button>}
+
```
Comment (LianaHus) [inline apps/doc-gen/src/app/App.tsx:38]: extra line

File: apps/doc-gen/src/app/docgen-client.ts [hunk 13]
```diff
@@ -0,0 +1,82 @@
...
+  async opendDocs(docs: string[]) {
+    await this.call('manager', 'activatePlugin', 'doc-viewer')
+    await this.call('tabs' as any, 'focus', 'doc-viewer')
+    await this.call('doc-viewer' as any, 'viewDocs', docs)
+  }
+
+  async generateDocs() {
+    this.docgen([this.build])
+  }
+
```
Comment (LianaHus) [inline apps/doc-gen/src/app/docgen-client.ts:81]: extra line

File: .gitignore [hunk 14]
```diff
@@ -60,4 +60,4 @@ testem.log
 libs/remix-node/
 libs/remix-niks/
 
-apps/remix-react
\ No newline at end of file
+apps/remix-react
```
Comment (bunsenstraat) [inline .gitignore:63]: sorry I added those test ignores to the file, these can be removed
libs/remix-node/
libs/remix-niks/

apps/remix-react

File: tsconfig.paths.json [hunk 15]
```diff
@@ -148,9 +148,12 @@
             "@remix-ui/solidity-uml-gen": [
                 "libs/remix-ui/solidity-uml-gen/src/index.ts"
             ],
+            "@remix-ui/docgenclient-ui": [
```
[the remark this answers is not in the document]
Comment (joeizang) [reply tsconfig.paths.json:151]: Ah..right you are. Thanks

File: apps/doc-gen/src/app/docgen-client.ts (hunk 1 above)
Comment (joeizang) [reply apps/doc-gen/src/app/docgen-client.ts:8]: right

File: .gitignore (hunk 14 above)
Comment (joeizang) [reply .gitignore:63]: ok

File: apps/doc-gen/src/app/docgen-client.ts (hunk 13 above)
Comment (joeizang) [reply apps/doc-gen/src/app/docgen-client.ts:81]: right

File: apps/doc-gen/src/app/docgen/common/properties.ts (hunk 11 above)
Comment (joeizang) [reply apps/doc-gen/src/app/docgen/common/properties.ts:13]: right. I think we should look into using prettier and eslint to auto format our code pages.

File: apps/doc-gen/src/app/docgen/site.ts (hunk 4 above)
Comment (joeizang) [reply apps/doc-gen/src/app/docgen/site.ts:122]: right

File: apps/doc-gen/src/app/docgen/templates.ts (hunk 5 above)
Comment (joeizang) [reply apps/doc-gen/src/app/docgen/templates.ts:88]: right

**Comment by joeizang** [comment]
> i was not able to make it work.
> 
> * both plugins are in plugin manager (I guess they shouldn't be there)
> * after activation one is empty the other one I don;t know what happens, I don't see it.
> * there is no Generate Docs in right click menu

Actually @LianaHus there isn't a generate doc in the context menu yet. It on the docgen plugin in vertical icons panel. There is a big button there that you can click

File: apps/doc-gen/project.json [hunk 16]
```diff
@@ -0,0 +1,59 @@
...
+  "targets": {
+    "build": {
+      "executor": "@nrwl/webpack:webpack",
+      "outputs": ["{options.outputPath}"],
+      "defaultConfiguration": "development",
+      "options": {
+        "compiler": "babel",
+        "outputPath": "dist/apps/doc-gen",
+        "index": "apps/doc-gen/src/index.html",
+        "baseHref": "/",
```
[the remark this answers is not in the document]
Comment (joeizang) [reply apps/doc-gen/project.json:17]: done

**Review by ryestew — APPROVED** [review]

**Pull request merged**